From af6f49120edcf3b1324c249954ba5b1dc2f02267 Mon Sep 17 00:00:00 2001
From: Istio Automation <istio-testing-bot@google.com>
Date: Thu, 3 Mar 2022 15:41:37 -0800
Subject: [release-1.12] gateway: avoid recreating Deployment on each kubectl
 apply (#37698)

* gateway: avoid recreating Deployment on each kubectl apply

Currently, kubectl apply annotation from Gateway is propogated to
Deployment which triggers a restart everytime we change any field.

* Update templates.go

Co-authored-by: John Howard <howardjohn@google.com>
---
 pilot/pkg/config/kube/gateway/templates.go        | 15 +++++++++++++++
 .../config/kube/gateway/templates/deployment.yaml |  4 ++--
 .../config/kube/gateway/templates/service.yaml    |  2 +-
 3 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/pilot/pkg/config/kube/gateway/templates.go b/pilot/pkg/config/kube/gateway/templates.go
index b5007fed45..72993576fe 100644
--- a/pilot/pkg/config/kube/gateway/templates.go
+++ b/pilot/pkg/config/kube/gateway/templates.go
@@ -52,6 +52,21 @@ func funcMap() template.FuncMap {
 		}
 		return strings.TrimSuffix(string(data), "\n")
 	}
+	f["omit"] = func(dict map[string]string, keys ...string) map[string]string {
+		res := map[string]string{}
+
+		omit := make(map[string]bool, len(keys))
+		for _, k := range keys {
+			omit[k] = true
+		}
+
+		for k, v := range dict {
+			if _, ok := omit[k]; !ok {
+				res[k] = v
+			}
+		}
+		return res
+	}
 	return f
 }
 
diff --git a/pilot/pkg/config/kube/gateway/templates/deployment.yaml b/pilot/pkg/config/kube/gateway/templates/deployment.yaml
index 2923c2cfeb..c3062f8a57 100644
--- a/pilot/pkg/config/kube/gateway/templates/deployment.yaml
+++ b/pilot/pkg/config/kube/gateway/templates/deployment.yaml
@@ -2,7 +2,7 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   annotations:
-    {{ toYamlMap .Annotations | nindent 4 }}
+    {{ toYamlMap (omit .Annotations "kubectl.kubernetes.io/last-applied-configuration") | nindent 4 }}
   labels:
     {{ toYamlMap .Labels
       (strdict "gateway.istio.io/managed" "istio.io-gateway-controller")
@@ -23,7 +23,7 @@ spec:
       annotations:
         {{ toYamlMap
           (strdict "inject.istio.io/templates" "gateway")
-          .Annotations
+          (omit .Annotations "kubectl.kubernetes.io/last-applied-configuration")
           | nindent 8}}
       labels:
         {{ toYamlMap
diff --git a/pilot/pkg/config/kube/gateway/templates/service.yaml b/pilot/pkg/config/kube/gateway/templates/service.yaml
index 31581c6dbc..d2e8af6646 100644
--- a/pilot/pkg/config/kube/gateway/templates/service.yaml
+++ b/pilot/pkg/config/kube/gateway/templates/service.yaml
@@ -2,7 +2,7 @@ apiVersion: v1
 kind: Service
 metadata:
   annotations:
-    {{ toYamlMap .Annotations | nindent 4 }}
+    {{ toYamlMap (omit .Annotations "kubectl.kubernetes.io/last-applied-configuration") | nindent 4 }}
   labels:
     {{ toYamlMap .Labels
       (strdict "gateway.istio.io/managed" "istio.io-gateway-controller")
-- 
2.31.1

